### PR TITLE
SI-7343 Fixed phase ordering in specialization

### PR DESCRIPTION
**Pull request by VladUreche**
Specialization rewires class parents during info transformation, and
the new info then guides the tree changes. But if a symbol is created
during duplication, which runs after specialization, its info is not
visited and thus the corresponding tree is not specialized.

One manifestation is the following:

```
object Test {

  class Parent[@specialized(Int) T]

  def spec_method[@specialized(Int) T](t: T, expectedXSuper: String) = {
     class X extends Parent[T]() 
     // even in the specialized variant, the local X class
     // doesn't extend Parent$mcI$sp, since its symbol has
     // been created after specialization and was not seen
     // by specialzation's info transformer.
     ...
  }
}
```

We can fix this by forcing duplication to take place before specialization.

Review by @dragos, @paulp or @axel22.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -1240,12 +1240,39 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
...
+     * }
+     * ```
+     * We fix this by forcing duplication to take place before specialization.
+     */
+    override def retyped(context: Context, tree: Tree, oldThis: Symbol, newThis: Symbol, env: scala.collection.Map[Symbol, Type]): Tree =
+      if (inSpecialization)
+        beforeSpecialize(super.retyped(context, tree, oldThis, newThis, env))
+      else // the constructors phase (which also uses duplication) comes after erasure and uses the
+           // post-erasure typer => we must protect it from the beforeSpecialization phase shifting.
+        super.retyped(context, tree, oldThis, newThis, env)
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: Tacking a Boolean on is one of my least favorite things; "super.retyped(context, tree, oldThis, newThis, env)" is not a string of characters which should occur twice in a method.

An alternative approach which involves less duplication and Boolean-tacking-on might go something like this:

```
override def retyped(context: Context, tree: Tree, oldThis: Symbol, newThis: Symbol, env: Map[Symbol, Type]): Tree =
  retypingWrapper(super.retyped(context, tree, oldThis, newThis, env))
protected def retypingWrapper[T](body: => T): T = body

// and at the instantiation site
val d = new specializeTypes.Duplicator(Map[Symbol, Type]()) {
  override protected def retypingWrapper[T](body: => T): T = beforeSpecialize(body)
}
```

**Comment by VladUreche** [comment]
Thanks for pointing it out Paul! Fixed it by subclassing `Duplicator` and shifting the phase there.

**Comment by paulp** [comment]
LGTM. That's better than my suggestion.

**Comment by gkossakowski** [comment]
@VladUreche: I wonder what was the effect when specialization was not rewiring the parent of the closure? Would you just get unspecialized closure and more boxing?

**Comment by VladUreche** [comment]
Yes, it inherited from the generic FunctionX and used the generic `apply`. If the info transformer isn't kicking in, the tree transformer won't touch the code, as it lacks the metadata to guide it.

Luckily (or unfortunately) this was limited to specialized bodies corresponding to methods, otherwise someone would have spotted the bug earlier.

**Comment by VladUreche** [comment]
Ping, can someone merge? The PR passes the merge criteria.

**Comment by VladUreche** [comment]
Thanks Adriaan!

**Comment by adriaanm** [comment]
Here's a variant that crashes the backend (replace val by var and accompanying assigment):

```
  class Val[@specialized(Int) T](t: T, expectedXSuper: String) {
    var check: T = _
    check = {
      class XXXXXX extends Parent[T]()
      // get the superclass for X and make sure it's correct
      val actualXSuper2 = (new XXXXXX).getClass().getSuperclass().getSimpleName()
      assert(actualXSuper2 == expectedXSuper, actualXSuper2 + " != " + expectedXSuper)
      t
    }
  }
```

```
warning: an unexpected type representation reached the compiler backend while compiling SI-7343.scala: => String#266. If possible, please file a bug on issues.scala-lang.org.
scala.MatchError: => String#266 (of class scala.reflect.internal.Types$NullaryMethodType)
    at scala.tools.nsc.backend.jvm.BTypesFromSymbols.typeToBType(BTypesFromSymbols.scala:208)
    at scala.tools.nsc.backend.jvm.BCodeHelpers$BCInnerClassGen$class.toTypeKind(BCodeHelpers.scala:355)
    at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.toTypeKind(BCodeSkelBuilder.scala:49)
    at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.symInfoTK(BCodeSkelBuilder.scala:79)
    at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:337)
    at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genEqEqPrimitive(BCodeBodyBuilder.scala:1265)
    at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genCond(BCodeBodyBuilder.scala:1196)
    at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genPrimitiveOp(BCodeBodyBuilder.scala:239)
    at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:663)
    at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:313)
```

**Comment by VladUreche** [comment]
@adriaanm, do you have a self-contained example I can try?

**Comment by adriaanm** [comment]
Sure! The example in my comment is derived from [the test case in this PR](https://github.com/scala/scala/pull/2651/files#diff-2b5042de633934e5e985ceb6f0befe17R37) by replacing the val by a var as illustrated.

**Comment by adriaanm** [comment]
(This is the new tree shape of valdefs between [fields](https://github.com/adriaanm/scala/tree/traits-late-fields) and constructors)

**Comment by VladUreche** [comment]
Thanks Adriaan, I can reproduce it now. It's a rather funny one, since just replacing `val` by `var` is not enough to trigger it. One has to do:

```
  var check: T = _
  check = { ... }
```

Does [your patch](https://github.com/adriaanm/scala/commit/854a45cab939515c5533e4eb1b9bd62cc579fbd4) fix the issue or should I have a look?

**Comment by adriaanm** [comment]
My patch triggers the bug because this is what valdefs now look like during specialization.

**Comment by VladUreche** [comment]
Got it, sorry for being slow today. I will fix it over the weekend and do a PR.

**Comment by VladUreche** [comment]
https://issues.scala-lang.org/browse/SI-9531

**Comment by adriaanm** [comment]
No worries -- thanks for looking into it!

**Comment by VladUreche** [comment]
@adriaanm, sorry for the lateness, I've been flooded with the spark summit talk. I will look into this on Friday.

**Comment by adriaanm** [comment]
No worries. Good luck with your talk.

**Comment by VladUreche** [comment]
@adriaanm, I performed a diagnosis of the problem, but I don't know what is the best way to fix it:
https://issues.scala-lang.org/browse/SI-9531?focusedCommentId=73572&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-73572

TLDR:

> Now, the easiest way to handle this would be to signal to the typer that we're inside a method, so it should treat values as local. Can we do this?
